test/unit/test_result_batch.py

```diff
@@ -0,0 +1,155 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+#
+# Copyright (c) 2012-2021 Snowflake Computing Inc. All right reserved.
+#
+
+import importlib
+from collections import namedtuple
+from http import HTTPStatus
+from unittest import mock
+
+import pytest
+from mock import Mock
+
+from snowflake.connector import DatabaseError, InterfaceError
+from snowflake.connector.compat import (
+    BAD_GATEWAY,
+    BAD_REQUEST,
+    FORBIDDEN,
+    GATEWAY_TIMEOUT,
+    INTERNAL_SERVER_ERROR,
+    METHOD_NOT_ALLOWED,
+    OK,
+    REQUEST_TIMEOUT,
+    SERVICE_UNAVAILABLE,
+    UNAUTHORIZED,
+)
+from snowflake.connector.errorcode import (
+    ER_FAILED_TO_CONNECT_TO_DB,
+    ER_FAILED_TO_REQUEST,
+)
+from snowflake.connector.errors import (
+    BadGatewayError,
+    BadRequest,
+    ForbiddenError,
+    GatewayTimeoutError,
+    InternalServerError,
+    MethodNotAllowed,
+    OtherHTTPRetryableError,
+    ServiceUnavailableError,
+)
+from snowflake.connector.result_batch import MAX_DOWNLOAD_RETRY, JSONResultBatch
+from snowflake.connector.sqlstate import (
+    SQLSTATE_CONNECTION_REJECTED,
+    SQLSTATE_CONNECTION_WAS_NOT_ESTABLISHED,
+)
+
+REQUEST_MODULE_PATH = (
+    "snowflake.connector.vendored.requests"
+    if importlib.util.find_spec("snowflake.connector.vendored.requests")
+    else "requests"
+)
+
+MockRemoteChunkInfo = namedtuple("MockRemoteChunkInfo", "url")
+
+
+def create_mock_response(status_code):
+    mock_resp = Mock()
+    mock_resp.status_code = status_code
+    mock_resp.raw = "success" if status_code == OK else "fail"
+    return mock_resp
+
+
+@mock.patch(REQUEST_MODULE_PATH + ".get")
+def test_ok_response_download(mock_get):
+    mock_get.return_value = create_mock_response(200)
+    chunk_info = MockRemoteChunkInfo("http://www.chunk-url.com")
+
+    result_batch = JSONResultBatch(100, None, chunk_info, [], [], True)
+    response = result_batch._download()
+
+    # successful on first try
+    assert mock_get.call_count == 1
+    assert response.status_code == 200
+
```

Comment:
The logic for the tests below are very similar but I kept them apart because they each test a different case